src/azul/openapi/validation.py

```diff
@@ -0,0 +1,242 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from copy import (
+    copy,
+)
+import json
+import logging
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    TypeVar,
+)
+
+import attr
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    RequirementError,
+)
+from azul.openapi.schema import (
+    python_type_for,
+)
+from azul.plugins.metadata.hca.transform import (
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, kw_only=False)
+class ValidationError(Exception):
+    message: str
+    name: str
+    value: Any
+    schema: Optional[JSON] = None
+    content: Optional[JSON] = None
+
+    def to_dict(self):
```

Comment:
attrs has stuff for this

Reply:
looking into this `attr.fields_dict` does not play nicely with exceptions.

Reply:
I think using a custom to_dict method is advantageous here, it allows us to remove inapplicable fields for a given parameter. For example, we don't need `contents=None` for a parameter that defines a `schema`.

Reply:
ah! can use attr.asdict() and pass in a filter 👍🏽